### PR TITLE
IPC module - initial Alpha

### PR DESCRIPTION
**Pull request by davidmiller**

### CODE REVIEW

File: plugins/ipc/lab.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    def __repr__(self):
+        return "\n".join(
+            [f"{self.test.lab_number} {self.test.test_code} {self.test.test_name} {self.test.datetime_ordered}"] +
+            [f"{o.observation_name} {o.observation_value}" for o in self.get_observations()]
+        )
+
+    def get_observations(self):
+        return self.test.observation_set.filter(
+            observation_name__in=self.OBSERVATION_NAMES
+        ).exclude(observation_value='.{}')
```
Comment (fredkingham) [inline plugins/ipc/lab.py:30]: I think you want to also exclude if observation = None as most of the `def alert` assume the existence of a string.

We have observations where the value is None, I'm guessing it should not happen but it _could_

Comment (davidmiller) [reply plugins/ipc/lab.py:30]: Observations of these codes and names  ? 

Surely if this was the case running `manage.py ipc` would stack trace?

File: plugins/ipc/lab.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    TEST_CODE         = 'DIFF'
+    OBSERVATION_NAMES = ['C.diff PCR test', 'C.diff Toxin test']
+    ALERT_CATEGORY    = models.InfectionAlert.CDIFF
+
+    def alert(self):
+        """
+        If either observation is Positive, return True
+        """
+        for observation in self.get_observations():
+            if observation.observation_value == 'POSITIVE':
```
Comment (fredkingham) [inline plugins/ipc/lab.py:68]: Are you sure we don't want observation.observation_value.startswith('POSITIVE')?

File: plugins/ipc/utils.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+def alphanum_key(s):
+    """ Turn a string into a list of string and number chunks.
+        "z23a" -> ["z", 23, "a"]
+    """
+    return [ tryint(c) for c in re.split('([0-9]+)', s) ]
+
+def sort_nicely(l):
+    """ Sort the given list in the way that humans expect.
+    """
+    return sorted(l, key=alphanum_key)
```
Comment (fredkingham) [inline plugins/ipc/utils.py:26]: This feels like it might break with an int str error, ie

a = ["z", 23, "a"]
b = [23, "z", "a"]
sorted([a, b]) -> raises an error

File: plugins/ipc/views.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        mrsas  = ['MRSA'] + [monthly[k][models.InfectionAlert.MRSA] for k in sorted(monthly.keys())[-12:]]
+        tbs    = ['TB'] + [monthly[k][models.InfectionAlert.TB] for k in sorted(monthly.keys())[-12:]]
+        vres   = ['VRE'] + [monthly[k][models.InfectionAlert.VRE] for k in sorted(monthly.keys())[-12:]]
+
+        context['overview_data'] = [ticks, cdiffs, cpes, mrsas, tbs, vres]
+
+        context['rfh_patients'] = UpstreamLocation.objects.filter(building='RFH').count()
+        context['rfh_siderooms'] = UpstreamLocation.objects.filter(building='RFH', room__startswith='SR').count()
+        context['weekly_alerts'] = models.InfectionAlert.objects.filter(
+            trigger_datetime__gte=today-datetime.timedelta(days=7)).count()
```
Comment (fredkingham) [inline plugins/ipc/views.py:45]: I may be reading this wrong, do we not just care about RFH Infection alerts?

File: plugins/admissions/management/commands/parse_admissions.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                room=encounter.pv1_3_room,
+                bed=encounter.pv1_3_bed
+            )
+            location.hospital  = encounter.pv1_3_hospital
+            location.patient   = encounter.patient
+            location.admitted  = encounter.pv1_44_admit_date_time
+            location.encounter = encounter
+
+            location.save()
+            self.stdout.write('.')
```
Comment (fredkingham) [inline plugins/admissions/management/commands/parse_admissions.py:44]: you probably don't need this

File: plugins/ipc/lab.py (hunk 2 above)
Comment (davidmiller) [reply plugins/ipc/lab.py:68]: Possibly, but we're reviewing this with clinical team later this week - right now these are indicative only

File: plugins/ipc/templates/records/infection_alert.html [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+Lab #: [[ item.lab_test_id ]]
+<span ng-show="item.seen">
+     Seen
+   <br />
+</span>
+<span ng-hide="item.seen">
+  Unseen
+</span>
+<span ng-show="item.comments">
+   <br />
```
Comment (fredkingham) [inline plugins/ipc/templates/records/infection_alert.html:11]: If a patient is seen they have a `<br>` after but if there are comments they have a `<br>` before, I think you just want `Unseen <br />`

File: plugins/ipc/utils.py (hunk 3 above)
Comment (davidmiller) [reply plugins/ipc/utils.py:26]: I was literally just copy and pasting Ned Batchelor https://nedbatchelder.com/blog/200712/human_sorting.html

Also, why do you think that raises an error? 

```
>>> a = ["z", 23, "a"]
>>> b = [23, "z", "a"]
>>> sorted([a, b])
[[23, 'z', 'a'], ['z', 23, 'a']]
>>>
```

File: plugins/ipc/views.py (hunk 4 above)
Comment (davidmiller) [reply plugins/ipc/views.py:45]: > do we not just care about RFH Infection alerts?

Can you rephrase the question?
I don't know what you're asking :)

File: plugins/ipc/templatetags/ipc.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+"""
+from django import template
+
+from plugins.admissions.models import UpstreamLocation
+
+from plugins.ipc.episode_categories import IPCEpisode
+
+register = template.Library()
+
+@register.inclusion_tag('templatetags/ward_count.html')
```
Comment (fredkingham) [inline plugins/ipc/templatetags/ipc.py:12]: This is probably a filter rather than a tag (doesn't really matter)
